Firestore/core/src/firebase/firestore/remote/grpc_stream.cc

```diff
@@ -206,6 +199,14 @@ void GrpcStream::MaybeUnregister() {
   }
 }
 
+void GrpcStream::FinishCall(const OnSuccess& callback) {
+  // All completions issued by this call must be taken off the queue before
+  // finish operation can be enqueued.
+  FastFinishCompletionsBlocking();
```

Comment:
This is in line to what gRPC folks described:
* calls should always be finished;
* all tags must be taken off the queue before Finish can be called.

The crash I saw happened when there were two calls to `GenericClientAsyncReaderWriter::Finish`. I'm not 100% sure, but one way it could happen is if network connectivity changed in-between `OnOperationFailed` and `OnFinishedByServer` (which is line to how I ran into this, trying to switch off network randomly):
* a stream has both a pending read and a pending write;
* one of those, say, the read, failed and triggered `OnOperationFailed`. At this point, the read is already removed from `completions_`, but the write is still there;
* `OnOperationFailed` calls `Finish` on the gRPC call. The write is on the completion queue;
* after this, but before the finish operation completes and invokes its completion handler, network connectivity changes, which invokes `FinishAndNotify` on this stream;
* `FinishAndNotify` invokes `Shutdown`, which notices `completions_` are not empty (there is still that write there), which it takes as a signal to call `Finish` on the gRPC call. However, this call has already been finished, which triggers an assertion failure in gRPC.

Reply:
I can see what you mean. In general performing all operations on the async queue should help alleviate the need for this kind of reasoning, but that's hard to do down here.

At the higher level we impose the restriction that callbacks are only honored if they were fired before a major state transition. Such a strategy doesn't work well for listeners registered to get callbacks indefinitely. An alternative is to ensure that routines called by the network monitor are idempotent.

If I'm reading correctly it seems that the latter is what you've done here. I'm down with that.

Reply:
Hmm, I think so. The new logic is that before `Finish` is invoked on the gRPC call, pending completions are always drained. Because `Finish` is a no-op when there are no completions, it effectively does make `Finish` idempotent.